auth-gateway/index.js

```diff
@@ -0,0 +1,162 @@
+require('dotenv').config();
+
+const polka = require('polka');
+const jwksRsa = require('jwks-rsa');
+const jwt = require('jsonwebtoken');
+
+const { AUTH0_DOMAIN, AUTH0_AUDIENCE } = process.env;
+const AUTH0_ISSUER = `https://${AUTH0_DOMAIN}/`;
+
+// Grabs public key, caches it, from Auth0's servers
+// asymmetric signature used to ensure that not only the jwt was unmodified
+// since signed, but that the issuer was Auth0 (since they hold the private key)
+const jwksFn = jwksRsa({
+  cache: true,
+  rateLimit: true,
+  jwksRequestsPerMinute: 5,
+  jwksUri: `https://${AUTH0_DOMAIN}/.well-known/jwks.json`
+});
+
+// Ref: https://github.com/auth0/node-jwks-rsa/blob/master/src/integrations/express.js
+function getSecretKey(header, cb) {
+  jwksFn.getSigningKey(header.kid, (err, key) => {
+    if (err) {
+      cb(err, null);
+      return;
+    }
+
+    cb(null, key.publicKey || key.rsaPublicKey);
+  });
+}
+
+// Verifies signature, as well as the issuer and audience, to guarantee
+// that the token was truly for our resource server
+const verifyToken = token =>
+  new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      getSecretKey,
+      {
+        issuer: AUTH0_ISSUER,
+        audience: AUTH0_AUDIENCE,
+        clockTolerance: 2, // seconds to deal with clock skew
+        algorithms: ['RS256']
+      },
+      (err, dToken) => {
+        if (err) {
+          console.error(err);
+          return reject(err);
+        }
+
+        return resolve(dToken);
+      }
+    );
+  });
+
+const bearerPrefix = 'Bearer ';
+const bearerPrefixLen = bearerPrefix.length;
+
+const cookiePrefix = 'access_token=';
+const cookieOffset = cookiePrefix.length;
+
+const getAuthToken = req => {
+  let whereFoundCount = 0;
+
+  let bearerIdx;
+  let cookieIdx;
+
```

Comment:
I'm OK with this function (modulo other comments) as is, but I'd like to share how I would approach it. I find a structure like the below easier to audit and review. In general, I try to avoid so-called "correlated conditionals" where there's a succession of `if`s. Reasoning about such a program is frustrated by the combinatorial explosion of possible execution paths (consider `if (a) ... else ... ; if (b) ... else ...` there's two conditions and 2^2 paths)

```javascript
let cookie_idx = req.headers.cookie ? req.headers.cookie.indexOf(cookiePrefix) : -1
let header_idx = req.headers.authorization ? req.headers.authorization.indexOf(bearerPrefix) : -1
if (cookie_idx > 0) {
    if (header_idx > 0) {
        console.warn('User specified > 1 access tokens', req.headers.origin);
        return null
    }
    return req.headers.cookie.substring(cookieIdx + cookieOffset);
} else if (header_idx > 0) {
    return req.headers.authorization.substring
} else {
    return null
}
```

In the case we must check the query parameter is not specified:
```javascript
let cookie_idx = req.headers.cookie ? req.headers.cookie.indexOf(cookiePrefix) : -1
let header_idx = req.headers.authorization ? req.headers.authorization.indexOf(bearerPrefix) : -1
let query_idx = // ...
if (cookie_idx > 0) {
    if (header_idx > 0 || query_idx > 0) {
        console.warn('User specified > 1 access tokens', req.headers.origin);
        return null
    }
    return req.headers.cookie.substring(cookieIdx + cookieOffset);
} else if (header_idx > 0) {
    if (query_idx > 0) {
        console.warn('User specified > 1 access tokens', req.headers.origin);
        return null
    }
    return req.headers.authorization.substring
} else {
    return null
}
```

Reply:
Sure, I don't see how our methods are much different, besides your use of ternary, which I agree gives terseness benefits.

A small point, in your statements above, I wouldn't write 

```js
if (cookie_idx > 0) {
    if (header_idx > 0 || query_idx > 0) {
        console.warn('User specified > 1 access tokens', req.headers.origin);
        return null
    }
    return req.headers.cookie.substring(cookieIdx + cookieOffset);
} else if (header_idx > 0) {
    if (query_idx > 0) {
        console.warn('User specified > 1 access tokens', req.headers.origin);
        return null
    }
    return req.headers.authorization.substring
} else {
    return null
}
```

`else` isn't needed after a return: https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style

I have, over time come to prefer no else after return. There are obviously people who disagree. Here's one: https://blog.mozilla.org/nnethercote/2009/08/31/no-else-after-return-considered-harmful/. I find that in practice long chains of if else if are much harder to read and offer no safety benefits.

However, if the Hail team wants `if{return '';}else{ }`, especially if there's some empirical evidence against the no else after return rule, I will change my approach.

Reply:
In python we've preferred the no-else style. Consistency across langs seems preferable, so I support the no-else style in JS as well.